pkg/rate/api_limiter.go

```diff
@@ -0,0 +1,790 @@
+	// params.RateBurst.
+	limiter *rate.Limiter
+
+	// currentRequestsInFlight is the number of parallel API requests
+	// currently in flight
+	currentRequestsInFlight int
+
+	// requestsProcessed is the total number of processed requests
+	requestsProcessed int64
+
+	// parallelWaitSemaphore is the semaphore used to implement
+	// params.MaxParallel. It is initialized with a capacity of
+	// waitSemaphoreResolution and each API request will acquire
+	// waitSemaphoreResolution/params.MaxParallel tokens.
+	parallelWaitSemaphore *semaphore.Weighted
+}
+
+// APILimiterParameters is the configuration of an APILimiter
+type APILimiterParameters struct {
+	// EstimatedProcessingDuration is the estimated duration an API call
+	// will take. This value is used if AutoAdjust is enabled to
+	// automatically adjust rate limits to stay as close as possible to the
+	// estimated processing duration.
+	EstimatedProcessingDuration time.Duration
+
+	// AutoAdjust enables automatic adjustment of the values
+	// ParallelRequests, RateLimit, and RateBurst in order to keep the
+	// mean processing duration close to EstimatedProcessingDuration
+	AutoAdjust bool
+
+	// MeanOver is the number of entries to keep in order to calculate the
+	// mean processing and wait duration
+	MeanOver int
+
+	// ParallelRequests is the parallel requests allowed. If AutoAdjust is
+	// enabled, the value will adjust automatically.
+	ParallelRequests int
+
+	// MaxParallelRequests is the maximum parallel requests allowed. If
+	// AutoAdjust is enabled, then the ParalelRequests will never grow
+	// above MaxParallelRequests.
+	MaxParallelRequests int
+
+	// MinParallelRequests is the minimum parallel requests allowed. If
+	// AutoAdjust is enabled, then the ParallelRequests will never fall
+	// below MinParallelRequests.
+	MinParallelRequests int
+
+	// RateLimit is the initial number of API requests allowed per second.
+	// If AutoAdjust is enabled, the value will adjust automatically.
+	RateLimit rate.Limit
+
+	// RateBurst is the initial allowed burst of API requests allowed. If
+	// AutoAdjust is enabled, the value will adjust automatically.
+	RateBurst int
+
+	// MinWaitDuration is the minimum time an API request always has to
+	// wait before the Wait() function returns an error.
+	MinWaitDuration time.Duration
+
+	// MaxWaitDuration is the maximum time an API request is allowed to
+	// wait before the Wait() function returns an error.
+	MaxWaitDuration time.Duration
+
+	// Log enables info logging of processed API requests. This should only
+	// be used for low frequency API calls.
+	Log bool
+
+	// DelayedAdjustmentFactor is percentage of the AdjustmentFactor to be
+	// applied to RateBurst and MaxWaitDuration defined as a value between
+	// 0.0..1.0. This is used to steer a slower reaction of the RateBurst
+	// and ParallelRequests compared to RateLimit.
+	DelayedAdjustmentFactor float64
+
+	// SkipInitial is the number of API calls to skip before applying rate
```

Comment:
Nitpick - this comment seems a bit misleading to me.

I assumed that "number of API calls to skip" meant that initial API calls would be rate limited ("skipped") and not that rate limiter doesn't kick in for them.

Maybe "SkipInitial is the number of API calls which will not have rate limits applied" would be better wording? 

This is not blocking at all.